### ISSUE TITLE
Keen Dreams doesn't start up

### ISSUE THREAD

**Issue by jschwartzenberg** [stsp/dosemu2#905]
**Describe the bug**
Keen Dreams doesn't start.

**To Reproduce**
1. Download the game: https://www.commander-keen.com/download-keen-dreams.php
2. Run it with `start`
3. Press 5
4. Observe, game hangs with "Please standby..."

**Attach the binaries**
https://www.commander-keen.com/download-keen-dreams.php

**Attach the log**
It is located in ~/.dosemu/boot.log
To make the log more useful, you may need to enable some logging flags.
See description of -D option in `man dosemu.bin`.

I have no clue how to make the log more useful :)
[boot.log](https://github.com/stsp/dosemu2/files/3494069/boot.log)

**Comment by jschwartzenberg** [comment]
The sources are here:
https://github.com/keendreams/keen

I should still try to compile them.

**Comment by stsp** [comment]
Does it start with freedos or with dosemu1?

**Comment by jschwartzenberg** [comment]
The same hang occurs in dosemu1 with freedos.

**Comment by jschwartzenberg** [comment]
I just managed to compile the sources. With my own build, I get the exact same hang.

**Comment by jschwartzenberg** [comment]
The problem is with the loop here: https://github.com/keendreams/keen/blob/4348b4719be87a0b873f5f4f77b4de82ddf7050c/id_sd.c#L204

When I disable the call to that function containing the loop, I can get to the menu. After that it crashes DOSEMU2 when I try to start a new game. The output of the crash is random. Attached just one example.
[boot.log](https://github.com/stsp/dosemu2/files/3506956/boot.log)

**Comment by stsp** [comment]
See here:
https://rawgit.com/stsp/dosemu2/devel/doc/EMUfailure.html#AEN32
You can change that calibration loop to use
the signed compare, and exit at `<=0`.

**Comment by jschwartzenberg** [comment]
I need a bit of either more info or more time to figure out how to do that ;)

But would that help with avoid the crash later on or is disabling the loop effectively the same? The Adlib/SB is detected and disabling sound doesn't avoid the crash either.

**Comment by stsp** [comment]
Use `0x7fff` instead of `0xffff` as initial vaue, and use for example
`test  [TimerDone],0x8000` to test for going below zero.

**Comment by jschwartzenberg** [comment]
I tried this, but that doesn't work:
```
-- a/id_sd.c
+++ b/id_sd.c
@@ -213,13 +213,13 @@ SDL_InitDelay(void)
        for (i = 0,timer = 0;i < 10;i++)        // Do timing test 10 times
        {
        asm     xor             dx,dx                                   // Zero DX
-       asm     mov             cx,0xffff                               // Put starting value in CX
+       asm     mov             cx,0x7fff                               // Put starting value in CX^M
        asm     mov             [TimerDone],cx                  // TimerDone = false - 1
 startloop:
        asm     or              [TimerDone],0
        asm     jnz             startloop                               // Make sure we're at the start
 loop:
-       asm     test    [TimerDone],1                   // See if TimerDone flag got hit
+       asm     test    [TimerDone],0x8000              // See if TimerDone flag got hit^M
        asm     jnz             done                                    // Yep - drop out of the loop
        asm     loop    loop
 done:
```

**Comment by stsp** [comment]
`        asm     or              [TimerDone],0`
Please alter that line, not the line you altered below that.
`jnz` can then be changed to `jz`

**Comment by jschwartzenberg** [comment]
That works!!
But after that the game still crashes after going through the menu: 
[boot.log](https://github.com/stsp/dosemu2/files/3510870/boot.log)

**Comment by jschwartzenberg** [comment]
There's a problem with my builds, they don't work in DOSBox either. I'll try different versions of Borland.

**Comment by jschwartzenberg** [comment]
As I type this, I got one of the versions to run... it plays fine apart from the scrolling issue.

**Comment by stsp** [comment]
So what was the problem?
Have you reported the crash upstream,
or have you submitted the timer fix to them?

**Comment by jschwartzenberg** [comment]
The crashes were probably due to a different version of Borland Turbo C++. Those executables didn't work properly in DOSBox either. I tried compiling different versions of the game until I found one that worked with Borland Turbo C++ 3.1. That turned out to be the 1.13 shareware version.

The upstream sources were primarily made available for historical preservation. I am not sure if there is a good place to report issues for the DOS version.
